core/math/aabb.cpp

```diff
@@ -159,7 +159,7 @@ bool AABB::find_intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, bool
 				}
 				tmax = t2;
 			}
-			if (tmin > tmax) {
+			if (tmin > tmax || (p_max_distance > 0 && tmin > p_max_distance)) {
```

Comment:
Are we now paying the price for this check even when not used? (`p_max_distance` defaulted to 0?)

I can see the `p_max_distance > 0` should be optimized out but does the second term `tmin > p_max_distance`?

Reply:
> Are we now paying the price for this check even when not used? (`p_max_distance` defaulted to 0?)
> 
> I can see the `p_max_distance > 0` should be optimized out but does the second term `tmin > p_max_distance`?

Yes we are. I believe nothing can be optimized out right now because the code is in the .cpp.
Will consider templating the function (therefore move it to the header file).

Reply:
Afaik compiler can optimize out functions in the cpp if called from the same translation unit (i.e. it has the opportunity to inline different versions of called functions), putting stuff in the header allows stuff to be optimized from a _different_ translation unit. (Link time optimization can theoretically also remove this limit but may not be as smart as in the same translation unit.)

Afaik usually this should mean wrapper functions in the same header / cpp will aggressively optimize.

Always worth checking the disassembly to see what is going on, or e.g. profile release version and watch the hot spots.